python/lsst/pipe/base/connectionTypes.py

```diff
@@ -249,10 +261,12 @@ class PrerequisiteInput(BaseInput):
       between dimensions are not desired (e.g. a task that wants all detectors
       in each visit for which the visit overlaps a tract, not just those where
       that detector+visit combination overlaps the tract).
+    - Prerequisite inputs may be optional (regular inputs are never optional).
```

Comment:
I'll clarify in the docs: for connections with `multiple=True`, `optional=True` means "a quantum may exist even if there are no datasets", while `optional=False` means "a quantum may exist if there are one or more inputs".

A `PipelineTask` can also override `adjustQuantum` to make their own cuts, but we cannot even make a quantum to pass to `adjustQuantum` if there are no regular `Input` datasets, and that's why they're always required.  With a new QG generation algorithm we might be able to relax that and allow `Input` connections to be optional, but I don't want to open that up early.

FWIW, I think this the only definition of `optional` + `multiple` that makes QG generation and execution consistent; if `optional=True and multiple=True` meant that some predicted-in-QG inputs may not materialize as available-at-execution inputs, then it means that QG generation time is special.  In other words, `PipelineTasks` with `multiple=True` connections cannot ever assume that what they see in QG gen is what they will see in execution, because they have to be able to accept any quantum that passes their `adjustQuantum`, regardless of when `adjustQuantum` was run.

Reply:
I dont think I like this at all, but it is going to take me some time to put it into words

Reply:
Let's discuss this live.  I'm free today until 1pm ET.